src/main/java/aztech/modern_industrialization/util/TransferHelper.java

```diff
@@ -67,32 +69,69 @@ public static void moveAll(IItemHandler src, IItemHandler target, boolean stackI
         }
     }
 
-    public static ItemStack extractMatching(IItemHandler src, Predicate<ItemStack> predicate, int maxAmount) {
-        int srcSlots = src.getSlots();
+    private interface ItemExtractor {
+        ItemStack get(int slot);
+
+        ItemStack extract(int slot, int amount);
+    }
 
+    private static ItemStack extractMatching(int srcSlots, ItemExtractor extractor, Predicate<ItemStack> predicate, int maxAmount,
```

Comment:
I don't like this, it's too complicated. Can you not do a first pass through the player inventory that will try to extract from the item-containing item and then a second pass calling the unmodified (except for bug fixes) `extractMatching`? As a side note, be careful with item-containing items of size > 1, it's a big mess and the handling is really ugly.

Reply:
Yes I can do that. It just means copying the old `extractMatching` and having duplicate code though, which isn't ideal. I'll make sure it excludes `IItemHandler` items with > 1 item, good point.

Reply:
Never mind that, I forgot `PlayerInvWrapper` exists.

Reply:
Hmm indeed. Duplication is not great either.

Reply:
But what is there to duplicate? What I have in mind is just to loop through every slot in `Inventory` and check if it contains a container item. If yes get the capability and call `extractMatching` on it. (Not sure we need to set the stack again; I don't think so).

Reply:
I've got what I think is a nifty solution, I'll push it in a few mins after I test it